### PR TITLE
[CircleEditor] Add Attributes Edit function

### PR DESCRIPTION
**Pull request by rudalsd**
This commit adds attributes edit function.
 -BuiltInOperator attribute value edit function.
 -CustomOperator attribute name, value and edit function.
 -CustomOperator attribute type edit function will be implemented.

Postmessage will be modified when the DS code is merged.

ONE-vscode-DCO-1.0-Signed-off-by: Gyeongmin Kwon rudalsd@naver.com

---

Related issue : #1269
Draft : #1298

### CODE REVIEW

File: media/CircleEditor/view-sidebar.js [hunk 1]
```diff
@@ -538,37 +603,261 @@ class NodeAttributeView {
...
+                // TODO Enable posting message
+                // vscode.postMessage({
+                //     command : 'alert',
+                //     text: 'Can\'t exceed 2,147,483,648'
+                // });
+                return;
+            }
+        }
+
+        // TODO It will be implemented when the attribute type of the CustomOperator can be
```
Comment (llFreetimell) [inline media/CircleEditor/view-sidebar.js:765]: Isn't it available to enable it?

File: media/CircleEditor/view-sidebar.js [hunk 2]
```diff
@@ -538,37 +603,261 @@ class NodeAttributeView {
...
-     * TODO Implement cancel()
-     * Attributes cancel function
-     */
+        const value = this._attribute.value;
 
-    /**
-     * TODO Implement remove()
-     * When in custom operator, attributes remove function
-     */
+        switch (type) {
```
Comment (llFreetimell) [inline media/CircleEditor/view-sidebar.js:650]: What is this switch-case statement for? @.@
I am hard to distinguish the purpose of each case...

File: media/CircleEditor/view-sidebar.js (hunk 1 above)
Comment (rudalsd) [reply media/CircleEditor/view-sidebar.js:765]: An error occurs because the function that receives the type of the custom operator's type is not implemented in the DS side yet.

File: media/CircleEditor/view-sidebar.js (hunk 2 above)
Comment (rudalsd) [reply media/CircleEditor/view-sidebar.js:650]: This code was already implemented in Netron, so I used it.

File: media/CircleEditor/view-sidebar.js [hunk 3]
```diff
@@ -538,37 +595,262 @@ class NodeAttributeView {
...
+            this._element.removeChild(this._element.lastChild);
+        }
+
+        if (!this._isCustom) {
+            this._editObject._attribute.name = this._attribute.name;
+            this._editObject._attribute._value = this._line.value;
+            this._editObject._attribute._type = this._attribute.type;
+        }
+        // } else {
+        //     this.makeEditCustomObject();
```
[the remark this answers is not in the document]
Comment (rudalsd) [reply media/CircleEditor/view-sidebar.js:794]: OK, I will move the function.

File: media/CircleEditor/view-sidebar.js (hunk 1 above)
Comment (llFreetimell) [reply media/CircleEditor/view-sidebar.js:765]: But you disabled posting a message and leave it as TODO....?

File: media/CircleEditor/view-sidebar.js (hunk 2 above)
Comment (llFreetimell) [reply media/CircleEditor/view-sidebar.js:650]: Thanks for the information :)
But it is my first time to see these codes and wonder what the purpose is...!
You might use these codes because you know the purpose of these codes, isn't it?

File: media/CircleEditor/view-sidebar.js [hunk 4]
```diff
@@ -538,37 +579,237 @@ class NodeAttributeView {
...
+            if (content && content.length > 1000) {
+                content = content.substring(0, 1000) + '\u2026';
+            }
+            if (content && typeof content === 'string') {
+                content = content.split('<').join('&lt;').split('>').join('&gt;');
+            }
+            let line;
+            console.log(typeName);
+            console.log(type);
+            console.log(typeName[type]);
```
Comment (llFreetimell) [inline media/CircleEditor/view-sidebar.js:637]: Please remove these

File: media/CircleEditor/view-sidebar.js [hunk 5]
```diff
@@ -538,37 +580,234 @@ class NodeAttributeView {
...
+        while (this._element.childElementCount) {
+            this._element.removeChild(this._element.lastChild);
+        }
+
+        if (!this._isCustom) {
+            this._editObject._attribute.name = this._attribute.name;
+            this._editObject._attribute._value = this._line.value;
+            this._editObject._attribute._type = this._attribute.type;
+        }
+        // } else {
```
Comment (llFreetimell) [inline media/CircleEditor/view-sidebar.js:747]: Why this is disabled?
This seems not related with message handling...

File: media/CircleEditor/view-sidebar.js [hunk 6]
```diff
@@ -538,37 +580,234 @@ class NodeAttributeView {
...
-     * TODO Implement remove()
-     * When in custom operator, attributes remove function
-     */
+        let content = new sidebar.Formatter(value, type).toString();
+        if (content && content.length > 1000) {
+            content = content.substring(0, 1000) + '\u2026';
+        }
+        if (content && typeof content === 'string') {
+            content = content.split('<').join('&lt;').split('>').join('&gt;');
+        }
```
Comment (llFreetimell) [inline media/CircleEditor/view-sidebar.js:633]: `content` is already `string` so `typeof content === string` seems redundant ...!

```suggestion
        if (content) {
            content = content.length > 1000 ? content.substring(0, 1000) + '\u2026' : content;
            content = content.split('<').join('&lt;').split('>').join('&gt;');
        }
```

File: media/CircleEditor/view-sidebar.js (hunk 5 above)
Comment (rudalsd) [reply media/CircleEditor/view-sidebar.js:747]: key.type is undefined because the function that receives the attribute type of the custom operator is not implemented in the DS side yet.

File: media/CircleEditor/view-sidebar.js [hunk 7]
```diff
@@ -42,6 +42,9 @@
 
 var sidebar = sidebar || {};
 var base = base || require('./external/base');
+var typeName = typeName || {};
```
[the remark this answers is not in the document]
Comment (rudalsd) [reply media/CircleEditor/view-sidebar.js]: we changed var to const, so we don't need to re-declare variable

File: media/CircleEditor/view-sidebar.js [hunk 8]
```diff
@@ -538,37 +575,232 @@ class NodeAttributeView {
...
 
-    /**
-     * TODO Implement edit()
-     * Attributes edit function
-     */
+    edit() {
+        while (this._element.childElementCount) {
+            this._element.removeChild(this._element.lastChild);
+        }
+        const type = this._attribute.type;
```
Comment (llFreetimell) [inline media/CircleEditor/view-sidebar.js:590]: `this._attribute.type` is used even `const type` is defined.
Either of them is fine, but just unify them :)

File: media/CircleEditor/type.js [hunk 9]
```diff
@@ -32,136 +32,3 @@ const tensorType =
     ['FLOAT32', 'FLOAT16', 'INT32', 'UINT8', 'INT64', 'BOOLEAN', 'INT16', 'INT8', 'FLOAT64'];
 
 const customType = ['int', 'boolean', 'string'];
-
-const builtinOperatorTypeEnum = {
```
[the remark this answers is not in the document]
Comment (rudalsd) [reply media/CircleEditor/type.js:36]: Yes, I left the reason at commit message.

**Review by llFreetimell — APPROVED** [review]
LGTM

**Comment by llFreetimell** [comment]
@rudalsd FYI, bullet `-` should have one empty space like `- blahblah`
